r/src/arrow_cpp11.h

```diff
@@ -157,8 +157,15 @@ struct ns {
 
 template <typename Pointer>
 Pointer r6_to_pointer(SEXP self) {
-  return reinterpret_cast<Pointer>(
-      R_ExternalPtrAddr(Rf_findVarInFrame(self, arrow::r::symbols::xp)));
+  void* p = R_ExternalPtrAddr(Rf_findVarInFrame(self, arrow::r::symbols::xp));
+  if (p == nullptr) {
```

Comment:
We can't really do that because of the implicit `else NULL` case: 

```r
shared_ptr <- function(class, xp) {
  if (!shared_ptr_is_null(xp)) class$new(xp)
}
```

There are cases where we want an R NULL when the internal C++ shared pointer holds a C++ null pointer, e.g. when we call: 

```c++
std::shared_ptr<Array> RecordBatch::GetColumnByName(const std::string& name) const {
  auto i = schema_->GetFieldIndex(name);
  return i == -1 ? NULLPTR : column(i);
}
```

For example in this tests: 

```r
test_that("[[ and $ on RecordBatch", {
  [...]
  expect_null(batch$qwerty)
  [...]
})
```

having the extra layer with the R function `shared_ptr(T, .)` maybe calling `T$new(.)` gives the NULL.

Reply:
Another way would be to have the ability to directly create the R6 objects internally, either by having more code in 

```cpp
template <typename T>
SEXP as_sexp(const std::shared_ptr<T>& ptr) {
  return cpp11::external_pointer<std::shared_ptr<T>>(new std::shared_ptr<T>(ptr));
}
```

but we would need some sort of dispatch from `T` to the R6 object. 

Or we would need to change the interface of many functions: 

```cpp
// [[arrow::export]]
std::shared_ptr<arrow::Array> StructArray__field(
    const std::shared_ptr<arrow::StructArray>& array, int i) {
  return array->field(i);
}
```

perhaps to: 

```cpp
// [[arrow::export]]
R6 StructArray__field(
    const std::shared_ptr<arrow::StructArray>& array, int i) {
  return R6(array->field(i), "Array");
}
```

but I don't think it's worth it

Reply:
Ah ok, I missed that that check translated some responses into `NULL`--though I don't think that's something we do widely, and maybe it would be better to make that more explicit in order to simplify the rest of the cases. 

Regarding directly returning R6 objects from the cpp code, my thought had been more like the first idea of modifying `as_sexp` to map `T` to R6 classes. Our convention is that R6 class names are generally the bare C++ class name (assuming namespaces), i.e. `arrow::dataset::Dataset` is `Dataset` R6 class, with some exceptions. So we could record a map of those exceptions (e.g. "arrow::csv::TableReader" maps to "CsvTableReader") in cpp. If I were writing this in R, it could look something like this:

```r
get_r6_constructor <- function(cpp_class) {
  if (cpp_class %in% names(special_cases)) {
    r6_name <- special_cases[[cpp_class]]
  } else {
    r6_name <- sub("^.*::(.*)$", "\\1", cpp_class)
  }
  get(r6_name, asNamespace("arrow"))$new
}
```

cc @bkietz

Reply:
I'd prefer to specify a complete mapping T to R6 class names, rather than only listing special cases. This would be a pretty straightforward trait structure in c++